proposals/nnnn-structured-concurrency.md

```diff
@@ -255,6 +257,148 @@ If the scope of an `async let` exits with a thrown error, the child task corresp
+
+```swift
+func chopVegetables(rawVeggies: [Vegetable]) async throws -> [ChoppedVegetable] {
+  Task.withNursery(resultType: ChoppedVegetable.self) { nursery in    
+    var choppedVeggies: [ChoppedVegetable] = []
+    choppedVeggies.reserveCapacity(veggies.count)
+        
+    // add all chopping tasks and process them concurrently
+    for v in rawVeggies {
+      await try nursery.add { // await the successful adding of the task 
+        await v.chopped() // await the processing result of task
+      }
+    }
+
+    while let choppedVeggie = await try nursery.next() { 
+      choppedVeggies.append(choppedVeggie)
+    }
+    
+    return choppedVeggies
+  }
+}
+```
+
+#### Nurseries: Throwing and cancellation
+
+Worth pointing out here is that adding a task to a nursery could fail because the nursery could have been cancelled when we were about to add more tasks to it. To visualize this, let us consider the following example:
+
+Tasks in a nursery by default handle thrown errors using like the musketeers would, that is: "*One for All, and All for One!*" In other words, if a single task throws an error, which escapes into the nursery, all other tasks will be cancelled and the nursery will re-throw this error.
+
+To visualize this, let us consider chopping vegetables again. One type veggetable that can be quite tricky to chop up is onions, they can make you cry if you don't watch out. If we attempt to chop up those vegetables, the onion will throw an error into the nursery, causing all other tasks to be cancelled automatically:
+
+```swift
+func chopOnionsAndCarrots(rawVeggies: [Vegetable]) async throws -> [Vegetable] {
+  await try Task.withNursery { nursery in // (3) will re-throw the onion chopping error
+    // kick off asynchronous vegetable chopping:
+    for v in rawVeggies {
+      await try nursery.add { 
+        await try v.chopped() // (1) throws
+      }
+    }
+    
+    // collect chopped up results:
+    while let choppedVeggie = await try nursery.next() { // (2) will throw for the onion
+      choppedVeggies.append(choppedVeggie)
+    }
+  }
+}
+```
+
+Let us break up the `chopOnionsAndCarrots()` function into multiple steps to fully understand its semantics:
+
+- first w add vegetable chopping tasks to the nursery
+- the chopping of the various vegetables beings asynchronously,
+- eventually an onion will be chopped and `throw`
+
+
+So far we did not yet discuss the cancellation of nurseries. A nursery can be cancelled form the outside, for example like this: 
+
+#### Nurseries: Implicitly awaited tasks
+Sometimes it is not necessary to gather the results of asynchronous functions (e.g. because they may be `Void` returning, "uni-directional"), in this case we can rely on the nursery implicitly awaiting for all tasks started before returning. 
```

Comment:
Well, these aren't the semantics I described, but they might be more reasonable.

Reply:
I see okey... The Task nurseries section in our "big doc" did not really have much detail and that's how I understood the 

  // Postcondition: if the body returns normally, the nursery is empty.
  // If it throws, all tasks in the nursery will be automatically cancelled.

Reply:
I checked around a bit to confirm my thinking here, the same seems to be the semantics exhibited by  Trio as well as Kotlin scopes. I think it's a very useful semantic to have, for reference:

- https://trio.readthedocs.io/en/stable/tutorial.html#okay-let-s-see-something-cool-already
- https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-scope/index.html#custom-usage

> By convention, they all wait for all the coroutines inside their block to complete before completing themselves, thus enforcing the structured concurrency. See Job documentation for more details.